rfcs/20240503-opset-deprecations.md

```diff
@@ -0,0 +1,102 @@
+# RFC: StableHLO v1.0 Opset Deprecations & Cleanups
+
+Author: gleasonk<br/>
+Last Modified: 5/3/24<br/>
+Status: In review<br/>
+
+## Background
+
+This doc covers a list of opset cleanups that we want to do for StableHLO v1.0.
+Most of these ops were never spec’ed and therefore have no formal compatibility
+guarantees, per the Unspecced Features compatibility exemption, however we can
+provide some backward / forward compatibility for most of them.
+
+This doc will propose futures for the ops intentionally omitted from the spec,
+including:
+
+- [“Not in HLO” ops][not-in-HLO] ([#3](https://github.com/openxla/stablehlo/issues/3)):
+`broadcast`, `create_token`, `cross-replica-sum`, `dot`, `einsum`,
+`torch_index_select`, `unary_einsum`, `trace` ([#604](https://github.com/openxla/stablehlo/issues/604)).
+- [Dynamism RFC P4](https://github.com/openxla/stablehlo/blob/main/rfcs/20230704-dynamism-101.md#p4)
+opset updates: `real_dynamic_slice` vs `dynamic_slice`.
+- Potentially unused ops like stateful `rng` [#597](https://github.com/openxla/stablehlo/issues/597)
+and `map`.
+
+In general (unless the op is unused and can be trivially deleted), the
+deprecation steps will be as follows:
+
+1. Migrate framework uses of redundant ops.
+1. Block serialization of deprecated ops once frameworks migrated.
+1. Migrate uses of the ops to the supported StableHLO op (add builder methods).
+1. Change VHLO legalization to upgrade to the supported op for compatibility.
+1. Remove the redundant StableHLO op.
+1. Remove redundant op from VHLO after 6mo.
+
+## Proposed Opset Changes
+
+### P0: Delete `CreateTokenOp` and `TraceOp`
+
+These ops are both unused as far as we can tell. They can be trivially deleted.
+
+### P1: Deprecate `BroadcastOp`, `DotOp`, `UnaryEinsumOp`
+
+These ops are all a trivial subset of features of another op. I.e. BroadcastOp
+can be represented using BroadcastInDim, DotOp with DotGeneralOp, UnaryEinsum
+with [`einsum` lowering][einsum-lowering].
+These ops will follow the formal deprecation process listed above.
+
+Helper methods can be added to the support op for compatibility, something like:
+`isLhsBroadcast`, `isSimpleDot`, `isUnaryEinsum`.
+
+### P2: Deprecate `RealDynamicSliceOp`, Enhance `DynamicSliceOp`
+
+In terms of naming `stablehlo.dynamic_slice` is more in-model than
+`real_final_v2_dynamic_slice`. However in terms of functionality, per
```

Comment:
Using the actual name of the op may be clearer, though I do think this is amusing.